packages/transaction-controller/src/utils/swaps.ts

```diff
@@ -0,0 +1,341 @@
+  // the `simulationFails` property here, we can reduce the number of swap
+  // transactions that get published to the blockchain only to fail and thereby
+  // waste the user's funds on gas.
+  if (
+    transactionType === TransactionType.swap &&
+    swaps?.hasApproveTx === false &&
+    transactionMeta.simulationFails
+  ) {
+    await cancelTransaction(transactionMeta.id);
+    throw new Error('Simulation failed');
+  }
+
+  const swapsMeta = swaps?.meta as Partial<TransactionMeta>;
+
+  if (!swapsMeta) {
+    return;
+  }
+
+  if (transactionType === TransactionType.swapApproval) {
+    updateSwapApprovalTransaction(transactionMeta, swapsMeta);
+    controllerHubEmitter(TransactionEvent.newSwapApproval, {
+      transactionMeta,
+    });
+  }
+
+  if (transactionType === TransactionType.swap) {
+    updateSwapTransaction(transactionMeta, swapsMeta);
+    controllerHubEmitter(TransactionEvent.newSwap, {
+      transactionMeta,
+    });
+  }
+}
+
+/**
+ * Attempts to update the post transaction balance of the provided transaction
+ *
+ * @param transactionMeta - Transaction meta object to update
+ * @param updatePostTransactionBalanceRequest - Dependency bag
+ * @param updatePostTransactionBalanceRequest.ethQuery - EthQuery object
+ * @param updatePostTransactionBalanceRequest.getTransaction - Reading function for the latest transaction state
+ * @param updatePostTransactionBalanceRequest.updateTransaction - Updating transaction function
+ */
+export async function updatePostTransactionBalance(
+  transactionMeta: TransactionMeta,
+  {
+    ethQuery,
+    getTransaction,
+    updateTransaction,
+  }: {
+    ethQuery: EthQuery;
+    getTransaction: (transactionId: string) => TransactionMeta | undefined;
+    updateTransaction: (transactionMeta: TransactionMeta, note: string) => void;
+  },
+) {
+  const transactionId = transactionMeta.id;
+
+  let latestTransactionMeta, approvalTransactionMeta;
+  for (let i = 0; i < UPDATE_POST_TX_BALANCE_ATTEMPTS; i++) {
+    const postTransactionBalance = await query(ethQuery, 'getBalance', [
+      transactionMeta.txParams.from,
+    ]);
+    latestTransactionMeta = getTransaction(transactionId) as TransactionMeta;
+    approvalTransactionMeta = latestTransactionMeta.approvalTxId
+      ? getTransaction(latestTransactionMeta.approvalTxId)
+      : null;
+
+    latestTransactionMeta.postTxBalance = postTransactionBalance.toString(16);
+    const isDefaultTokenAddress = isSwapsDefaultTokenAddress(
+      transactionMeta.destinationTokenAddress as string,
+      transactionMeta.chainId,
+    );
+
+    if (
+      isDefaultTokenAddress &&
+      transactionMeta.preTxBalance === latestTransactionMeta.postTxBalance
+    ) {
+      await sleep(UPDATE_POST_TX_BALANCE_TIMEOUT);
+    } else {
+      break;
+    }
+  }
+
+  updateTransaction(
+    latestTransactionMeta as TransactionMeta,
+    'TransactionController#updatePostTransactionBalance - Add post transaction balance',
+  );
+  return Promise.resolve({
```

Comment:
Minor, but no need to use `Promise.resolve` here since this is now an `async` method.